Public/Src/Cache/ContentStore/Library/Service/Grpc/BufferedWriteStream.cs

```diff
@@ -0,0 +1,112 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace BuildXL.Cache.ContentStore.Service.Grpc
+{
+    // A write-only stream backed by a fixed-size buffer that
+    // processes content when the buffer is full and then
+    // re-uses the buffer.
+
+    // This class is required for serving compressed data via grpc
+    // because the compression api insists on writing to a stream and
+    // the grpc api exposes an write-one-chunk method instead of a stream.
+
+    // This class is not safe for simultaneous use by multiple threads.
+
+    // Be sure to call Flush when done writing, otherwise the last bit
+    // of content will (probably) not be processed, because it will (probably)
+    // not fill the buffer.
+    internal class BufferedWriteStream : Stream
+    {
+        public BufferedWriteStream(byte[] storage, Func<byte[], int, int, Task> writer)
+        {
+            if (storage is null) { throw new ArgumentNullException(nameof(storage)); }
+            if (writer is null) { throw new ArgumentNullException(nameof(writer)); }
+            _storage = storage;
+            _writer = writer;
+            _writePointer = 0;
+            _position = 0;
+        }
+
+        private readonly byte[] _storage; // buffer to store content before written
+        private readonly Func<byte[], int, int, Task> _writer; // method to write
+        private int _writePointer; // pointer to next index in buffer to be written to
+        private int _position; // total bytes written
+
+        public override bool CanWrite => true;
+
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cts)

```

Comment:
nit: Should it be 'cancellationToken' instead of cts?

(I think 'cts' usually means 'CancellationToken*Source*'.

Reply:
Fixed